pkg/recommender/test/recommender_test.go

```diff
@@ -0,0 +1,562 @@
+			queryExpr:      queryWillReturnFalse,
+			expectedResult: false,
+		},
+		{
+			name:           "Incorrect query expression",
+			rangedQueryFun: SimpleRangedQueryFunction(),
+			queryExpr:      incorrectQueryExpr,
+			errorExpected:  true,
+		},
+		{
+			name: "Ranged query returns arbitrary error",
+			rangedQueryFun: func(string, v1.Range) (model.Value, v1.Warnings, error) {
+				return nil, v1.Warnings{}, errors.New("Arbitrary error")
+			},
+			errorExpected: true,
+		},
+		{
+			name: "Ranged query returns unexpected value type",
+			rangedQueryFun: func(string, v1.Range) (model.Value, v1.Warnings, error) {
+				return model.Vector{}, v1.Warnings{}, nil
+			},
+			errorExpected: true,
+		},
+		{
+			name: "Query returns empty result Matrix",
+			rangedQueryFun: func(string, v1.Range) (model.Value, v1.Warnings, error) {
+				return model.Matrix{}, v1.Warnings{}, nil
+			},
+			errorExpected: true,
+		},
+		{
+			name: "Query returns Matrix containing empty record",
+			rangedQueryFun: func(string, v1.Range) (model.Value, v1.Warnings, error) {
+				return model.Matrix{{}}, v1.Warnings{}, nil
+			},
+			errorExpected: true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			metrics.Qr = test.rangedQueryFun
+			if test.queryExpr == "" {
+				test.queryExpr = queryWillReturnTrue
+			}
+			res, err := r.ExportedRangedQueryOnMetricsProvider(ctx, test.queryExpr, test.duration, test.argStep)
+
+			if !test.errorExpected {
+				if err != nil {
+					t.Errorf("test \"%s\" error, err %v", test.name, err)
+				} else if test.expectedResult != res {
+					t.Errorf("test \"%s\" expected result %v, got %v", test.name, test.expectedResult, res)
+				}
+			}
+		})
+	}
+}
+
+func TestGetRackRecommendations(t *testing.T) {
+	const (
+		rackName          = "rack_0_name"
+		ruleName          = "rule_0_name"
+		baseMembers       = 3
+		baseCpu           = "5"
+		memory            = "1Gi"
+		minAllowedMembers = 1
+		maxAllowedMembers = 100
+	)
+	var (
+		minAllowedCpu = resource.MustParse("1")
+		maxAllowedCpu = resource.MustParse("100")
+	)
+	ctx := log.WithNewTraceID(context.Background())
+	atom := zap.NewAtomicLevelAt(zapcore.InfoLevel)
+	logger, _ := log.NewProduction(log.Config{
+		Level: atom,
+	})
+	r := recommender.GetEmptyRecommender(logger)
+	r.SetMetricsProviderFakeAPI(metrics.MockApi{})
+	metrics.Qr = SimpleRangedQueryFunction()
+	metrics.Q = SimpleQueryFunction()
+
+	tests := []struct {
+		name           string
+		rack           *scyllav1.RackSpec
+		scalingPolicy  *v1alpha1.RackScalingPolicy
+		expectedResult *v1alpha1.RackRecommendations
+		errorExpected  bool
+	}{
+		{
+			name: "Recommend scaling members",
+			rack: getRackSpec(rackName, baseMembers, baseCpu, baseCpu, memory, memory),
+			scalingPolicy: getScalingPolicy(rackName,
+				[]v1alpha1.ScalingRule{
+					*getScalingRule(ruleName, 1, queryWillReturnTrue, nil, nil, v1alpha1.ScalingModeHorizontal, 2.0),
```

Comment:
give a name to these magic numbers

Reply:
Ok. Added:
`
const (
		rackName          = "rack_name"
		ruleName          = "rule_name"
		baseMembers       = 3
		baseCpu           = "5"
		higherCpu = "40"
		memory            = "1Gi"
		minAllowedMembers = 1
		maxAllowedMembers = 100
		priority1 = 1
		priority2 = 2
		priority3 = 3
		factor2 = 2.0
		factor4 = 4.0
		factor6 = 6.0
	)
	var (
		minAllowedCpu = resource.MustParse("1")
		maxAllowedCpu = resource.MustParse("100")
		duration5 = util.DurationPtr(5)
		duration10 = util.DurationPtr(10)
	)
	`
	and, if test results in scaling cpu, used it as follows:
	`corev1.ResourceCPU:    resource.MustParse(stringMulFloat64(baseCpu, factor4)),
`